app/controllers/oauths_controller.rb

```diff
@@ -0,0 +1,31 @@
+class OauthsController < ApplicationController
+  skip_before_action :require_login
+
+  def oauth
+    session[:return_to_url] = request.referer unless request.referer =~ /oauth/
+    login_at(params[:provider])
+  end
+
+  def callback
+    provider = params[:provider]
+    if @user = login_from(provider)
+      flash[:notice] = "Успешный вход через #{provider.titleize }"
+      redirect_to root_path
+    else
+      begin
+        @user = create_from(provider)
+        reset_session
+        auto_login(@user)
+        redirect_to root_path
+      rescue
+        redirect_to root_path, alert: 'Failed LogIn'
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.